### ISSUE TITLE
Allow an embedded Flutter View on Android to size itself

### ISSUE THREAD

**Issue by jiahaog** [flutter/flutter#149033]
This is like https://github.com/flutter/flutter/issues/134501 and https://github.com/flutter/flutter/issues/169147, but for Android.

A use case for addtoapp would be to embed Flutter into a native scrollable of some sort. Consider the following:

![image](https://github.com/flutter/flutter/assets/7111741/f8cea4e5-2322-4805-90f2-32b2815f07a7)

At this time, on mobile platforms, the size of the Flutter view must be fixed in the embedding. This reduces the design space for the embedded Flutter app.

```xml
<LinearLayout
  android:id="@+id/flutter_view_container"
  android:layout_width="match_parent"
  android:layout_height="500dp" />
```

Instead, we want to be able to set `wrap_content`:

```xml
<LinearLayout
  android:id="@+id/flutter_view_container"
  android:layout_width="match_parent"
  android:layout_height="wrap_content"/>
```

See b/342534475 for more internal context.

cc @xster

**Comment by jonahwilliams** [comment]
I don't believe the framework has any concept of a window with constraints, it needs exact tight bounds. I'm not sure how you would add this, from looking at the contents its unclear how the bounds negotiation would actually take place.

**Comment by jonahwilliams** [comment]
What information would a FlutterView use to size itself? Normally we just have exact tight bounds on a window. What information can the android embedding give to the flutter view?

**Comment by jiahaog** [comment]
Apologies, but I'm not familiar with the specifics around this. Maybe @goderbauer can help here?

I was also filing this bug because this was mentioned in [Flutter’s multi-window status (PUBLICLY SHARED)
](https://docs.google.com/document/d/13E27tD8_9f6lDgwg3MpGNTV8XIRCZH3ByI-t9kI9IUM/edit#heading=h.zdkpxbn6ohaq) but isn't tracked:

> ### Content-sized views
> Windows that need to be sized to their content (e.g. Tooltips or menus) can be implemented on top of the recently added [dynamically sized view](https://github.com/flutter/flutter/pull/138648) feature: the window is configured with unconstrained constraints and the view size calculated from the content is passed to `FlutterView.render`.
> Each platform needs to be updated to resize the window before rendering the content. Currently, this is only implemented on Flutter web.

**Comment by goderbauer** [comment]
> I don't believe the framework has any concept of a window with constraints, it needs exact tight bounds. I'm not sure how you would add this, from looking at the contents its unclear how the bounds negotiation would actually take place.

To allow dynamically sized `FlutterView`s on the web (and for eventually sizing windows based on content) we added the following APIs:
* The engine can give a `FlutterView` `ViewConstraints` in `FlutterView.physicalConstraints`. Those constraints (like `BoxConstraints` for RenderObjects) can define a max/min height/width for the view. Currently, most platforms just specify tight constraints where max=min.
* The framework uses those view constraints as input for the layout algorithm (they basically become the `BoxConstraints` for the root `RenderObject`). We run the regular constraints-in size-out layout algorithm for the render tree, to come up with the size for the root RenderObject that is within the provided view constraints.
* When it is time to send the current `Scene` to the engine, the framework calls `FlutterView.render` and provides it with the  computed size with the expectation that the FlutterView is sized accordingly before showing the scene.

There is one important caveat here: We have a decent amount of widgets that currently do not work in an environment where they don't get tight constraints. So, they cannot be used unless an ancestor gives them tight constraints.

Relevant API docs:
* https://main-api.flutter.dev/flutter/dart-ui/FlutterView/physicalConstraints.html
* https://main-api.flutter.dev/flutter/dart-ui/ViewConstraints-class.html
* https://main-api.flutter.dev/flutter/dart-ui/FlutterView/render.html

**Comment by jacobsimionato** [comment]
I'd like to figure out the best process to go through to try to get this feature prioritized for a specific client engagement (see internal bug b/416093153), and also to improve the quality of the overall mobile add-to-app offering.

Can someone on the engine/framework side roughly estimate the size of this project? Would it need to be considered at the quarterly planning level, or is it a smaller piece of work that could be prioritized at a lower level?

The overall goal here would be to follow what https://github.com/flutter/flutter/issues/134501 achieved on web. E.g. allow the developer to express constraints for the overall Flutter view, and then have the native container that holds Flutter be resized based on Flutter content.

**Comment by jmagman** [comment]
I'm going to split this issue into two, since there will be separate work on iOS and Android.  This issue will be repurposed to track the Android side of the work, and the iOS side will be https://github.com/flutter/flutter/issues/169147.

**Comment by goderbauer** [comment]
> Is there a list of widgets that do support non-tight constraints. I am currently doing some testing and am unsure if I am using one such widget.

I am not aware of such a list - and the only way I am aware of to find these widgets is trial and error. If you are giving your app non-tight physical constraints and everything is working well then you are not using any such widget. If a widget doesn't support it, it will blow up in a very obvious way. We did fix the Overlay widget in https://github.com/flutter/flutter/pull/139513, which was the most prominent problematic widget that you'd constantly run into.

**Comment by mboetger** [comment]
[PR WIP](https://github.com/flutter/flutter/pull/176063)
